File: crypto/src/benchmark/cplusplus/blacklemon.cpp
/*
 * Copyright (c) 2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

#include <benchmark/benchmark.h>

#include "blacklemon.h"
#include "fastrng.h"

using namespace blacknet::crypto;

static FastDRG rng;

static void BM_BlackLemon_GenerateSecretKey(benchmark::State& state) {
    BlackLemon bl;
    BlackLemon::SecretKey sk;

    for (auto _ : state) {
        sk = bl.generateSecretKey(rng);

        benchmark::DoNotOptimize(sk);
        benchmark::DoNotOptimize(rng);
        benchmark::ClobberMemory();
    }
}
BENCHMARK(BM_BlackLemon_GenerateSecretKey);

static void BM_BlackLemon_GeneratePublicKey(benchmark::State& state) {
    BlackLemon bl;
    BlackLemon::SecretKey sk = bl.generateSecretKey(rng);
    BlackLemon::PublicKey pk;

    for (auto _ : state) {
        pk = bl.generatePublicKey(rng, sk);

        benchmark::DoNotOptimize(sk);
        benchmark::DoNotOptimize(pk);
        benchmark::DoNotOptimize(rng);
        benchmark::ClobberMemory();
    }
}
BENCHMARK(BM_BlackLemon_GeneratePublicKey);

static void BM_BlackLemon_Encrypt(benchmark::State& state) {
    BlackLemon bl;
    BlackLemon::SecretKey sk = bl.generateSecretKey(rng);
    BlackLemon::PublicKey pk = bl.generatePublicKey(rng, sk);
    BlackLemon::PlainText pt = BlackLemon::Rt::random(rng);
    BlackLemon::CipherText ct;

    for (auto _ : state) {
        ct = bl.encrypt(rng, pk, pt);

        benchmark::DoNotOptimize(pk);
        benchmark::DoNotOptimize(pt);
        benchmark::DoNotOptimize(ct);
        benchmark::DoNotOptimize(rng);
        benchmark::ClobberMemory();
    }
}
BENCHMARK(BM_BlackLemon_Encrypt);

static void BM_BlackLemon_Decrypt(benchmark::State& state) {
    BlackLemon bl;
    BlackLemon::SecretKey sk = bl.generateSecretKey(rng);
    BlackLemon::CipherText ct = { BlackLemon::Rq::random(rng), BlackLemon::Rq::random(rng)};
    BlackLemon::PlainText pt;

    for (auto _ : state) {
        pt = bl.decrypt(sk, ct);

        benchmark::DoNotOptimize(sk);
        benchmark::DoNotOptimize(pt);
        benchmark::DoNotOptimize(ct);
        benchmark::ClobberMemory();
    }
}
BENCHMARK(BM_BlackLemon_Decrypt);

File: crypto/src/main/rust/blacklemon.rs
/*
 * Copyright (c) 2024-2026 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use crate::algebra::{IntegerRing, One, PolynomialRing, Zero};
use crate::lpr;
use crate::random::UniformGenerator;

// https://blacknet.ninja/blacklemon.pdf

const KAPPA: usize = 2;
#[expect(dead_code)]
const ELL: usize = lpr::D;
const R: <lpr::Zq as IntegerRing>::Int = 40;

pub struct SecretKey {
    a: lpr::SecretKey,
    b: lpr::RqNTT,
}

pub struct PublicKey {
    a: lpr::PublicKey,
    b: lpr::RqNTT,
}

pub type CipherText = lpr::CipherText;

pub type PlainText = lpr::PlainText;

pub fn generate_secret_key<RNG: UniformGenerator<Output = u8>>(rng: &mut RNG) -> SecretKey {
    SecretKey {
        a: lpr::generate_secret_key(rng),
        b: lpr::generate_uniform(rng),
    }
}

pub fn generate_public_key<RNG: UniformGenerator<Output = u8>>(
    rng: &mut RNG,
    sk: &SecretKey,
) -> PublicKey {
    PublicKey {
        a: lpr::generate_public_key(rng, &sk.a),
        b: -sk.b,
    }
}

pub fn encrypt<RNG: UniformGenerator<Output = u8>>(
    rng: &mut RNG,
    pk: &PublicKey,
    pt: &PlainText,
) -> CipherText {
    let mut ct = lpr::encrypt(rng, &pk.a, pt);
    ct.a += pk.b;
    ct
}

pub fn decrypt(sk: &SecretKey, ct: &CipherText) -> PlainText {
    let ct = CipherText {
        a: ct.a + sk.b,
        b: ct.b,
    };
    lpr::decrypt(&sk.a, &ct)
}

pub fn detect(sk: &SecretKey, ct: &CipherText) -> Option<PlainText> {
    let mut m = lpr::Rt::default();
    let d = ct.a + ct.b * sk.a.s + sk.b;
    let coefficients = d.coefficients();
    for i in 0..lpr::D {
        if coefficients[i].absolute() <= R {
            m[i] = lpr::Zt::ZERO;
        } else if lpr::DELTA - coefficients[i].absolute() <= R {
            m[i] = lpr::Zt::ONE;
        } else {
            return None;
        }
    }
    if m.coefficients()
        .into_iter()
        .take(KAPPA)
        .any(|coefficient| coefficient != lpr::Zt::ZERO)
    {
        return None;
    }
    Some(PlainText { m })
}
